README.md
# tailRisk
A C++ library for the calculation of various tail risk measures

## Documentation
**NB: Documentation still under development.**

* Implemented [Risk Measures](RiskMeasures.md)
* Mathematical Documentation available [here](https://www.openriskmanual.org/wiki/Category:Tail_Risk) 


## Dependencies

* Armadillo (Data container and calculation library)
* Poco (For parsing JSON inputs)
* Stats (A C++ header-only library of statistical distribution functions.)

## Example
The data directory contains sample datafiles with various sampled distributions

```c++
    // Read in some data for a type 0 representation (discrete distribution)
int LossGrid = 1000;
int DataType = 0;
RandomVar L(LossGrid, DataType);
L.ReadFromJSON("../data/example5.json");
L.Print();

// Calculate various measures
double alpha = 0.8;
int threshold = 0;

std::cout << "Mean Value: " << L.Mean() << std::endl;
std::cout << "Median Value: " << L.Median() << std::endl;
std::cout << "STD Value: " << L.StandardDeviation() << std::endl;
std::cout << "Kurtosis: " << L.Kurtosis() << std::endl;
std::cout << "Skeweness: " << L.Skeweness() << std::endl;
std::cout << "Quantile @ " << alpha << ": " << L.Quantile(alpha) << std::endl;
std::cout << "Quantile Index @ " << alpha << ": " << L.Quantile_Index(alpha) << std::endl;
std::cout << "VaR @ " << alpha << ": " << L.VaR(alpha) << std::endl;
std::cout << "Expected Shortfall @ " << alpha << ": " << L.ExpectedShortFall(alpha) << std::endl;
std::cout << "Exceedance Probability: " << L.ExceedanceProbability(threshold) << std::endl;
std::cout << "Mean Excess: " << L.MeanExcess(threshold ) << std::endl;
```

ChangeLog.md
# TailRisk ChangeLog

* Nov 20, 2022, switch to eigen containers
* Nov 9, 2020, first public release

testing/test_vectors.cpp
/*################################################################################
  ##
  ##   Copyright (C) 2020-2022 Open Risk (www.openriskmanagement.com)
  ##
  ##   This file is part of the tailRisk C++ library.
  ##
  ##   Licensed under the Apache License, Version 2.0 (the "License");
  ##   you may not use this file except in compliance with the License.
  ##   You may obtain a copy of the License at
  ##
  ##       http://www.apache.org/licenses/LICENSE-2.0
  ##
  ##   Unless required by applicable law or agreed to in writing, software
  ##   distributed under the License is distributed on an "AS IS" BASIS,
  ##   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  ##   See the License for the specific language governing permissions and
  ##   limitations under the License.
  ##
  ################################################################################*/


#include <catch2/catch_test_macros.hpp>

TEST_CASE("vectors can be sized and resized", "[e-vector]") {

    std::vector<int> v(5);

    REQUIRE(v.size() == 5);
    REQUIRE(v.capacity() >= 5);

    SECTION("resizing bigger changes size and capacity") {
        v.resize(10);

        REQUIRE(v.size() == 10);
        REQUIRE(v.capacity() >= 10);
    }SECTION("resizing smaller changes size but not capacity") {
        v.resize(0);

        REQUIRE(v.size() == 0);
        REQUIRE(v.capacity() >= 5);
    }SECTION("reserving bigger changes capacity but not size") {
        v.reserve(10);

        REQUIRE(v.size() == 5);
        REQUIRE(v.capacity() >= 10);
    }SECTION("reserving smaller does not change size or capacity") {
        v.reserve(0);

        REQUIRE(v.size() == 5);
        REQUIRE(v.capacity() >= 5);
    }
}

testing/test_factorial.cpp
/*################################################################################
  ##
  ##   Copyright (C) 2020-2022 Open Risk (www.openriskmanagement.com)
  ##
  ##   This file is part of the tailRisk C++ library.
  ##
  ##   Licensed under the Apache License, Version 2.0 (the "License");
  ##   you may not use this file except in compliance with the License.
  ##   You may obtain a copy of the License at
  ##
  ##       http://www.apache.org/licenses/LICENSE-2.0
  ##
  ##   Unless required by applicable law or agreed to in writing, software
  ##   distributed under the License is distributed on an "AS IS" BASIS,
  ##   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  ##   See the License for the specific language governing permissions and
  ##   limitations under the License.
  ##
  ################################################################################*/


// This tells Catch to provide a main() - only do this in one cpp file
#include <catch2/catch_test_macros.hpp>

unsigned int Factorial(unsigned int number) {
    return number <= 1 ? number : Factorial(number - 1) * number;
}

TEST_CASE("Factorials are computed", "[b-factorial]") {
    REQUIRE(Factorial(1) == 1);
    REQUIRE(Factorial(2) == 2);
    REQUIRE(Factorial(3) == 6);
    REQUIRE(Factorial(10) == 3628800);
}

testing.cpp
/*################################################################################
  ##
  ##   Copyright (C) 2020-2022 Open Risk (www.openriskmanagement.com)
  ##
  ##   This file is part of the tailRisk C++ library.
  ##
  ##   Licensed under the Apache License, Version 2.0 (the "License");
  ##   you may not use this file except in compliance with the License.
  ##   You may obtain a copy of the License at
  ##
  ##       http://www.apache.org/licenses/LICENSE-2.0
  ##
  ##   Unless required by applicable law or agreed to in writing, software
  ##   distributed under the License is distributed on an "AS IS" BASIS,
  ##   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  ##   See the License for the specific language governing permissions and
  ##   limitations under the License.
  ##
  ################################################################################*/


/**
 * File: testing.cpp
 * Date: Apr 16 2021
 */

#include "stats.hpp"
#include "random_var.h"

int main(int argc, char *argv[]) {

    int SampleSize = 100;
    int DataType = 1;
    RandomVar myR(SampleSize, DataType);

    std::cout << "Testing" << std::endl;
    for (int i = 0; i < myR.size(); i++) {
        double rval = stats::rt(30);
        myR.setS(i, rval);
    }

    myR.Print();

}


random_var.h
/*################################################################################
  ##
  ##   Copyright (C) 2020-2022 Open Risk (www.openriskmanagement.com)
  ##
  ##   This file is part of the tailRisk C++ library.
  ##
  ##   Licensed under the Apache License, Version 2.0 (the "License");
  ##   you may not use this file except in compliance with the License.
  ##   You may obtain a copy of the License at
  ##
  ##       http://www.apache.org/licenses/LICENSE-2.0
  ##
  ##   Unless required by applicable law or agreed to in writing, software
  ##   distributed under the License is distributed on an "AS IS" BASIS,
  ##   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  ##   See the License for the specific language governing permissions and
  ##   limitations under the License.
  ##
  ################################################################################*/


/**
 * File:   random_var.h
 * Date: Mon Nov  9 10:31:35 CET 2020
 */

#ifndef RANDOM_VARIABLE_H
#define RANDOM_VARIABLE_H

#include <iostream>
#include <Eigen/Dense>

class RandomVar {
    friend std::ostream &operator<<(std::ostream &os, const RandomVar &R);

public:

    // constructor with size and empirical distribution type
    // Type 0 -> histogram storage
    // Type 1 -> sample storage

    RandomVar(size_t S, int type) {
        if (type == 0) {
            m_type = 0;
            m_P.resize(S);
            m_C.resize(S);
            m_X.resize(S);
            m_size = S;
        } else if (type == 1) {
            m_type = 1;
            m_S.resize(S);
            m_size = S;
        } else {
            std::cout << "Error in random variable representation type" << std::endl;
        }
    }

    // constructor directly from existing data
    // TODO generalize data type to accommodate different accuracy requirements

    RandomVar(Eigen::ArrayXd x, Eigen::ArrayXd p, const int size) {
        m_type = 0;
        m_size = size;
        m_X.resize(size);
        m_P.resize(size);
        m_C.resize(size);
        for (int i = 0; i < size; i++) {
            m_X[i] = x[i];
            m_P[i] = p[i];
        }
    }

    // overload assignment operator
    RandomVar &operator=(const RandomVar &R);

    /**
     *
     * @return
     */
    [[nodiscard]] size_t size() const {
        return m_size;
    };

    [[nodiscard]] int type() const {
        return m_type;
    };

    [[nodiscard]] double getP(int index) const {
        return m_P[index];
    };

    [[nodiscard]] double getC(int index) const {
        return m_C[index];
    };

    [[nodiscard]] double getX(int index) const {
        return m_X[index];
    };

    [[nodiscard]] double getS(int index) const {
        return m_S[index];
    };

    [[nodiscard]] double Average();

    [[nodiscard]] double Mean();

    [[nodiscard]] double Median();

    [[nodiscard]] double Variance();

    [[nodiscard]] double Vol();

    [[nodiscard]] double StandardDeviation();

    [[nodiscard]] double Kurtosis();

    [[nodiscard]] double Skeweness();

    [[nodiscard]] double ExpectedShortFall(double alpha);

    [[nodiscard]] double ExceedanceProbability(int index);

    [[nodiscard]] double MeanExcess(int index);

    [[nodiscard]] int Quantile_Index(double alpha);

    [[nodiscard]] double Quantile(double alpha);

    [[nodiscard]] double VaR(double alpha);


    void setP(int index, double arg) {
        m_P[index] = arg;
    };

    void setC(int index, double arg) {
        m_C[index] = arg;
    };

    void setX(int index, double arg) {
        m_X[index] = arg;
    };

    void setS(int index, double arg) {
        m_S[index] = arg;
    };

    void Sort();

    void Cumulative();

    void Probability();

    void ReadFromJSON(const std::string &filename);

    void Print();

private:
    // 0 Type: exact representation
    // 1 Type: sampling representation
    int m_type;

    int m_size;

    bool m_sorted = false;

    Eigen::ArrayXd m_P{}; // storage of probability mass
    Eigen::ArrayXd m_C{}; // storage of cumulative probability
    Eigen::ArrayXd m_X{}; // storage of discrete values (random variable range)
    Eigen::ArrayXd m_S{}; // storage of sampling data from simulation experiments
};

#endif


random_var.cpp
/*################################################################################
 ##
 ##  Copyright (C) 2020-2022 Open Risk (www.openriskmanagement.com)
 ##
 ##  This file is part of the tailRisk C++ library.
 ##
 ##  Licensed under the Apache License, Version 2.0 (the "License");
 ##  you may not use this file except in compliance with the License.
 ##  You may obtain a copy of the License at
 ##
 ##    http://www.apache.org/licenses/LICENSE-2.0
 ##
 ##  Unless required by applicable law or agreed to in writing, software
 ##  distributed under the License is distributed on an "AS IS" BASIS,
 ##  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 ##  See the License for the specific language governing permissions and
 ##  limitations under the License.
 ##
 ################################################################################*/

/**
 * File: random_var.cpp
 * Date: Mon Nov 9 10:31:35 CET 2020
 */

#include <cmath>
#include <iostream>
#include <cassert>
#include <fstream>


#include <Poco/JSON/Parser.h>
#include "random_var.h"

using namespace Poco;

/**
 * Assignment Operator
 * @param R
 * @return
 */
RandomVar &RandomVar::operator=(const RandomVar &R) {
    assert(R.size() == this->size()); // check that size matches
    for (size_t i = 0; i < R.size(); i++) {
        this->setX(i, R.getX(i));
        this->setP(i, R.getP(i));
        this->setC(i, R.getC(i));
    }
    return (*this);
};

/**
 * Sort the sampling data
 */
void RandomVar::Sort() {
    if (m_sorted == false) {
        std::sort(m_S.begin(), m_S.end());
    }
    m_sorted = true;
}

/**
 * Compute the cumulative distribution function
 */
void RandomVar::Cumulative() {
    m_C[0] = m_P[0]; // starting value
    for (size_t i = 1; i < m_P.size(); i++)
        m_C[i] = m_C[i - 1] + m_P[i]; // add probability mass
}

/**
 * Compute the probability density
 */
void RandomVar::Probability() {
    m_P[0] = m_C[0]; // starting valued
    for (size_t i = 1; i < m_P.size(); i++)
        m_P[i] = m_C[i] - m_C[i - 1]; // incremental probability mass
}

/**
 * Compute the average value
 * The calculation depends on how the data are stored
 */
double RandomVar::Average() {
    double expectation = 0.0;
    // from probability mass
    if (m_type == 0) {
        for (size_t i = 0; i < m_P.size(); i++) {
            expectation += m_P[i] * m_X[i];
        }
    } // from samples
    else if (m_type == 1) {
        for (size_t i = 0; i < m_S.size(); ++i) {
            expectation += m_S[i];
        }
        expectation /= m_S.size();
    }
    return expectation;
}

/**
 * Synonym for Average
 */
double RandomVar::Mean() {
    return Average();
}

/**
 * Median as Quantile(0.5)
 */
double RandomVar::Median() {
    return Quantile(0.5);
}

/**
 * Compute the variance
 * The calculation depends on how the data are stored
 */
double RandomVar::Variance() {
    double var = 0;
    double mean = Average();
    if (m_type == 0) {
        for (size_t i = 0; i < m_P.size(); i++)
            var += m_P[i] * m_X[i] * m_X[i];
        var -= mean * mean;
    } else if (m_type == 1) {
        for (size_t i = 0; i < m_S.size(); i++)
            var += (m_S[i] - mean) * (m_S[i] - mean);
        var /= (m_S.size() - 1) ;
    }
    return var;
}

/**
 * Compute volatility (standard deviation)
 */
double RandomVar::Vol() {
    return sqrt(Variance());
}

double RandomVar::StandardDeviation() {
    return Vol();
}

/**
 * Compute the skeweness
 * The calculation depends on how the data are stored
 */
// TODO Type 1 calculation
double RandomVar::Skeweness() {
    double skew = 0;
    if (m_type == 0) {
        double mean = Average();
        for (size_t i = 0; i < m_P.size(); i++)
            skew += m_P[i] * pow(m_X[i] - mean, 3);
        skew = skew / pow(Variance(), 3 / 2);
    }
    return skew;
}

/**
 * Compute the kurtosis
 * The calculation depends on how the data are stored
 */
// TODO Type 1 calculation
double RandomVar::Kurtosis() {
    double kurt = 0;
    if (m_type == 0) {
        double mean = Average();
        for (size_t i = 0; i < m_P.size(); i++)
            kurt += m_P[i] * pow(m_X[i] - mean, 4);
        kurt = kurt / pow(Variance(), 2);
    }
    return kurt;
}

/**
 * Compute the Quantile Index
 * The calculation depends on how the data are stored
 */
// TODO Type 1 calculation
int RandomVar::Quantile_Index(double alpha) {
    int index = 0;
    if (m_type == 0) {
        for (int i = 0; i < m_P.size(); i++) {
            if (m_C[i] > 1 - alpha) {
                index = i;
                break;
            }
        }
    } else if (m_type == 1) {
        this->Sort();
        index =  alpha * this->size();
    }
    return index;
}

/**
 * Compute the Quantile
 * The calculation depends on how the data are stored
 */
// TODO Type 1 calculation
double RandomVar::Quantile(double alpha) {
    double q;
    int index = this->Quantile_Index(alpha);
    if (m_type == 0) {
        q = m_X[index];
    }
    else if (m_type == 1){
        q = m_S[index];
    }
    return q;
}

/**
 * Compute Value-at-Risk
 */
double RandomVar::VaR(double alpha) {
    return Quantile(1.0 - alpha);
}

/**
 * Compute Expected Shortfall
 */
// TODO Type 1 calculation
double RandomVar::ExpectedShortFall(double alpha) {
    double es = 0;
    if (m_type == 0) {
        int iVaR = this->Quantile_Index(alpha);
        for (int k = iVaR; k < m_P.size(); k++) {
            es += m_P[k] * m_X[k];
        }
        es /= alpha;
    }
    return es;
}

/**
 * Compute Expeedance Probability
 */
// TODO Type 1 calculation
double RandomVar::ExceedanceProbability(int index) {
    double ep = 0;
    if (m_type == 0) {
        for (size_t k = index; k < m_P.size(); k++)
            ep += m_P[k];
    }
    return ep;
}

/**
 * Compute Mean Excess
 */
// TODO Type 1 calculation
double RandomVar::MeanExcess(int index) {
    double mx = 0;
    if (m_type == 0) {
        double alpha = ExceedanceProbability(index);
        for (size_t k = index; k < m_P.size(); k++)
            mx += m_P[k] * m_X[k];
        mx /= alpha;
    }
    return mx;
}

/**
 * overload << operator
 */
std::ostream &operator<<(std::ostream &os, const RandomVar &R) {
    std::ostringstream out;
    for (size_t k = 0; k < R.size(); k++)
        out << R.getX(k) << "\t" << R.getP(k) << "\t" << R.getC(k) << std::endl;
    return os << out.str();
};

/**
 * Print a representation
 */
void RandomVar::Print() {
    if (this->m_type == 1) {
        for (size_t s = 0; s < this->m_S.size(); s++) {
            std::cout << s << "\t" << this->m_S[s] << std::endl;
        }
    } else if (this->m_type == 0) {
        for (size_t s = 0; s < this->m_X.size(); s++) {
            std::cout << s << "\t" << this->m_X[s] << "\t" << this->m_P[s] << "\t" << this->m_C[s] << std::endl;
        }
    }
}

/**
 * Read random variable from JSON
 * @param fileName
 */
void RandomVar::ReadFromJSON(const std::string &filename) {

    // Random variable data are an array of objects
    //[
    // {"value": 1, "probability" : 0.2, "cumulative" : 0.2},
    // {"value": 2, "probability" : 0.2, "cumulative" : 0.4},
    // {"value": 3, "probability" : 0.2, "cumulative" : 0.6},
    // {"value": 4, "probability" : 0.2, "cumulative" : 0.8},
    // {"value": 5, "probability" : 0.2, "cumulative" : 1.0}
    //]

    Poco::JSON::Parser Parser;
    const char *cstr = filename.c_str();
    std::ifstream t(cstr);
    std::stringstream buffer;
    buffer << t.rdbuf();
    std::string json = buffer.str();
    // Parse the JSON and get the Results
    Parser.parse(json);
    Poco::Dynamic::Var ParsedJsonResult = Parser.result();

    Poco::JSON::Array::Ptr arr = ParsedJsonResult.extract<Poco::JSON::Array::Ptr>();
    size_t size = arr->size();
    std::cout << "Reading " << size << " records." << std::endl;

    m_P.resize(size);
    m_C.resize(size);
    m_X.resize(size);

    // Individual data rows
    Poco::JSON::Object::Ptr object;
    for (size_t i = 0; i < size; i++) {
        object = arr->getObject(i);
        this->setX(i, object->getValue<double>("value"));
        this->setP(i, object->getValue<double>("probability"));
        this->setC(i, object->getValue<double>("cumulative"));
    }
}


main.cpp
/*################################################################################
  ##
  ##   Copyright (C) 2020-2022 Open Risk (www.openriskmanagement.com)
  ##
  ##   This file is part of the tailRisk C++ library.
  ##
  ##   Licensed under the Apache License, Version 2.0 (the "License");
  ##   you may not use this file except in compliance with the License.
  ##   You may obtain a copy of the License at
  ##
  ##       http://www.apache.org/licenses/LICENSE-2.0
  ##
  ##   Unless required by applicable law or agreed to in writing, software
  ##   distributed under the License is distributed on an "AS IS" BASIS,
  ##   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  ##   See the License for the specific language governing permissions and
  ##   limitations under the License.
  ##
  ################################################################################*/

/**
 * File: main.cpp
 * Date: Mon Nov  9 10:31:35 CET 2020
 *
 * Examples of using the tailRisk library to compute tail risk measures
 *
 */

#include "random_var.h"

int main(int argc, char *argv[]) {

    // Reading in some data for a type 0 representation (discrete distribution)
    int LossGrid = 1000;
    int DataType = 0;
    RandomVar L(LossGrid, DataType);
    L.ReadFromJSON("../../data/example5.json");
    L.Print();

    // Calculate various measures
    double alpha = 0.8;
    int threshold = 0;

    std::cout << "Mean Value: " << L.Mean() << std::endl;
    std::cout << "Median Value: " << L.Median() << std::endl;
    std::cout << "STD Value: " << L.StandardDeviation() << std::endl;
    std::cout << "Kurtosis: " << L.Kurtosis() << std::endl;
    std::cout << "Skeweness: " << L.Skeweness() << std::endl;
    std::cout << "Quantile @ " << alpha << ": " << L.Quantile(alpha) << std::endl;
    std::cout << "Quantile Index @ " << alpha << ": " << L.Quantile_Index(alpha) << std::endl;
    std::cout << "VaR @ " << alpha << ": " << L.VaR(alpha) << std::endl;
    std::cout << "Expected Shortfall @ " << alpha << ": " << L.ExpectedShortFall(alpha) << std::endl;
    std::cout << "Exceedance Probability: " << L.ExceedanceProbability(threshold) << std::endl;
    std::cout << "Mean Excess: " << L.MeanExcess(threshold ) << std::endl;

    return 0;
}